### ISSUE TITLE
New function suggestions.

### ISSUE THREAD

**Issue by siraisisatoru** [alexlibby/hexo-tag-webp#1]
Hello, I was trying to rebuild my blog from Jekyll. However, Hexo does not provide proper JPG to WEBP format converter neither official nor 3rd parties. It causes me to panic for weeks until I found this 'plugin'. Before everything, I am new to JS, HEXO, and Github. So please forgive me for any non-sense explanation.

Base on the ```index.js``` in the most current versions, it will convert all PNG files into WEBP files.  However, it will convert all for every time you insert a new tag in the post. By referencing the process of [jekyll_picture_tag](https://github.com/rbuchberger/jekyll_picture_tag/tree/v1.14.0) plugin, I tried to develop a detection mechanism via changing the output path. Additionally, some image preprocesses and cropping function has been added.

I wonder it will be proper to leave a comment to provide information here or not, I will leave a brief description here. 

===================================================

The ```index.js``` now becomes:
``` js
const imagemin = require("imagemin-keep-folder");
const webp = require("imagemin-webp");
const path = require("path");
const webpconfig = hexo.config.webpconfig;
const fs = require("fs");
const md5 = require("js-md5");
const imageminJpegAutorotate = require("./imagemin-jpeg-autorotate");
const imageDir = "./source/image";
var sizeOf = require("image-size");

var webpSettings = {
  quality: 70,
};

function getimagFunc(args) {
  const originPic = args[0];
  const dimensions = sizeOf("./source/" + originPic);

  // console.log(dimensions);
  webpSettings.width =
    dimensions.orientation == 1 ? dimensions.width : dimensions.height;
  webpSettings.height =
    dimensions.orientation == 1 ? dimensions.height : dimensions.width;

  const inputImageAspectRatio = webpSettings.width / webpSettings.height;

  var imagFunc = {
    method: 6,
    quality: webpSettings.quality,
  };

  if (args[1]) {
    if (args[1].indexOf(":") < 0) {
      // first element is not ratio => self and the following arg is weight and height
      var picDimension = [args[1], args[2]];
      imagFunc.resize = { width: picDimension[0], height: picDimension[1] };
    } else {
      var picRatio = args[1].split(":").map(Number);
      picRatio = picRatio[0] / picRatio[1];
      let ratioW = webpSettings.width;
      let ratioH = webpSettings.height;
      if (inputImageAspectRatio > picRatio) {
        ratioW = webpSettings.height * picRatio;
      } else if (inputImageAspectRatio < picRatio) {
        ratioH = webpSettings.width / picRatio;
      }
      const shiftArr = [
        (webpSettings.width - ratioW) * 0.5,
        (webpSettings.height - ratioH) * 0.5,
      ];

      imagFunc.crop = {
        x: shiftArr[0],
        y: shiftArr[1],
        width: ratioW,
        height: ratioH,
      };
    }
  }
  if (args[2] && args[3]) {
    var picDimension = [args[2], args[3]];
    imagFunc.resize = { width: picDimension[0], height: picDimension[1] };
  }

  return imagFunc;
}

function picProcess(args) {
  const originPic = path.parse(args[0]);

  var argsPar = [...args];
  argsPar.shift();
  argsPar = argsPar.join(" ");
  const parMD5 = md5.hex(argsPar).slice(0, 9);

  const oFilename = "./source" + args[0];

  var moveToP = "./source/image" + originPic.dir + "/";

  moveToP = moveToP.replace("/post_image", "") + parMD5 + "/";
  const finalP = moveToP + originPic.name + ".webp";

  if (!fs.existsSync(finalP)) {
    if (!fs.existsSync(moveToP)) {
      fs.mkdirSync(moveToP, { recursive: true });
      console.log("Directory is created. - " + moveToP);
    }
    const imagFunc = getimagFunc(args);
    imagemin([oFilename], moveToP, {
      plugins: [
        imageminJpegAutorotate({
          disable: false,
        }),
        webp(imagFunc),
      ],
    }).then(() => {
      console.log("Images optimized - " + originPic.name + ".webp");
    });
  }

  return `
  <img src=${finalP.replace("./source", "")} class="nofancybox"/>
  `;
}

hexo.extend.tag.register("webp", function (args) {
  return picProcess(args);
});

```

What it does are the following:

1. ratio
``` js
{% webp /post_image/ABC/IMG_0639.JPG 6:9 %}
 ```
The output image will be cropped with the coresposing ratio

2. ratio and post-process width and height
``` js
{% webp /post_image/ABC/IMG_0638.JPG 16:9 300 200 %} 
```
The output image will be cropped with the corresponding ratio and then resize the image to corresponding size (width 300 height 200)

3. resize process width and height
``` js
{% webp /post_image/ABC/IMG_0640.JPG 300 200 %} 
```
Pure resize the image to the corresponding size (width 300 height 200)

===================================================

New files needed:
Please refer to the [issue for applying auto-rotate to input JPG](https://github.com/imagemin/imagemin/issues/229#issuecomment-402008994) to have added two files ```imagemin-jpeg-autorotate.js``` and ```jpeg-autorotate.js```. 

===================================================

My version for extensions:
```
    "image-size": "^1.0.0",
    "imagemin": "^8.0.0",
    "imagemin-keep-folder": "^5.3.2",
    "imagemin-webp": "^6.0.0",
    "is-jpg": "^2.0.0",
    "jpeg-autorotate": "^7.1.1",
    "js-md5": "^0.7.3"
```

===================================================

New file structure:
Now you can place your images into a folder (any name inside the root source folder) and then the output images will be generated into a sub-folder ```source/image```.

    .
    ├── ...
    ├── source
    │   ├── yourImageSourceFolder
    │   │   └── IMG_0001.JPG
    │   └── image
    │       └── 7890abcde
    │           └── IMG_0001.webp
    └── ...

The short HEX index is generated by the input configuration.

===================================================

Remarks:
1. The code I provided can be used without notification. 
2. The input can be any type of image that imagemin plugin supports (in theory, maybe not since jpeg-autorotate is used. I don't use other types than JPG, so I may not help with that.)
3. In my site is using pure WEBP format for all images. if you need to include the original image as well, please refer to the original ```index.js``` in this repo and add by yourself.
4. The cropping function will crop the middle part of the image, so you can modify the function to accept an offset parameter on your own.
5. Apology for my poor English and the long issue here. Hope this helps newcomers.
6. If you want to use this in the layout templates, please add the following code below the above code.
7. I hard code the parameters, you can also change them by refer to the original ```index.js``` in the repo.

```js
hexo.extend.helper.register("webp_helper", function (args) {
  return  picProcess(args);
});
```

Call the function by 

```js
 <%- webp_helper( '/post_image/ABC/IMG_0001.JPG 16:9' ) %>
```

**Comment by alexlibby** [comment]
Hi  @siraisisatoru - apologies for not responding sooner: it's been very busy for me over the last few weeks!

Thankyou for this code - it is very detailed, and has some great ideas! I actually wrote the original plugin as part of a book I wrote on Hexo, called "Practical Hexo". Ideally I would like to keep the plugin aligned to the book if I can, so it matches what is in the book. However, I do like what I see in your suggestions, so maybe there is something we can do as a newer version? Please let me consider what needs to be done in more detail - I will let you know what happens as soon as I can.

**Comment by siraisisatoru** [comment]
Hello @alexlibby. Thx for the reply.  For me, this plugin provided a guideline to begin but obviously, I don't have the ability to modify it into a universal image minimization plugin. Sometime before, I tried to use the modified version for a PNG file and seems not doing the conversion for me. There are still have many potential problems with supporting multiple image formats. 
I leave my code here for newcomers and provide a much clear guideline to do modifications. As it is originally related to your book, I don't think adding new functions to it is a great idea seems it will cause in-continuity to the readers. I suggest adding a new branch to this repo and mention to README file is more proper. Another option is to leave it is and this issue shows examples to the newcomers how to do modifications which is the simplest.
The regeneration of compressed images obviously is a bug and it is pretty ignored. 
As you are the owner of this repo, I leave the decision to you.

**Comment by siraisisatoru** [comment]
Hello. I keep working on building my blog for months. One day I need to add watermarks to my image but imagemin does not have a plugin for that and imagemin is now officially unmaintained. Therefore, I decided to change from imagemin to sharp.
The script I used in my blog is published in this [repo](https://github.com/siraisisatoru/hexo-sharp-script). If one day newcomers came here, my script can be one of the alternative options.
